src/algo_tx.c

```diff
@@ -164,6 +164,122 @@ map_kv_bin(uint8_t **p, uint8_t *e, char *key, uint8_t *valbuf, int vallen)
   return 1;
 }
 
+static int
+map_kv_accts(uint8_t **p, uint8_t *e, char *key, uint8_t accounts[][32], size_t num_accounts)
+{
+  if (num_accounts == 0) {
+    return 0;
+  }
+
+  if (num_accounts > FIXARR_15 - FIXARR_0) {
+    os_sched_exit(0);
+  }
+
+  encode_str(p, e, key, SIZE_MAX);
+
+  uint8_t *arrbase = *p;
+  if (*p >= e) {
+    // We need to access arrbase[0] below, so if there isn't space for at least
+    // one byte, bail out.
+    return 0;
+  }
+
+  put_byte(p, e, FIXARR_0);
+  for (size_t i = 0; i < num_accounts; i++) {
+    encode_bin(p, e, accounts[i], sizeof(accounts[0]));
+    arrbase[0]++;
+  }
+
+  return 1;
+}
+
+static int
+map_kv_args(uint8_t **p, uint8_t *e, char *key, uint8_t app_args[][MAX_ARGLEN], size_t app_args_len[], size_t num_app_args)
+{
+  if (num_app_args == 0) {
+    return 0;
+  }
+
+  if (num_app_args > FIXARR_15 - FIXARR_0) {
+    os_sched_exit(0);
+  }
+
+  encode_str(p, e, key, SIZE_MAX);
+
+  uint8_t *arrbase = *p;
+  if (*p >= e) {
+    // We need to access arrbase[0] below, so if there isn't space for at least
+    // one byte, bail out.
+    return 0;
+  }
+
+  put_byte(p, e, FIXARR_0);
+  for (size_t i = 0; i < num_app_args; i++) {
+    encode_bin(p, e, app_args[i], app_args_len[i]);
```

Comment:
Could we combine `map_kv_accts` and `map_kv_args` and `map_kv_fnapps` into some kind of `map_kv_array` that takes as a callback a function that will do `encode_bin(p, e, app_args[i], app_args_len[i]);` or `encode_uint64(p, e, foreign_apps[i]);` or `encode_bin(p, e, accounts[i], sizeof(accounts[0]));`?